Sources/Fuzzilli/FuzzIL/AbstractInterpreter.swift

```diff
@@ -35,83 +126,28 @@ public struct AbstractInterpreter {
-            typeChanges += getStateChanges(oldState: functionState, newState: stack.last!)
         case is BeginIf:
-            stack.append(currentState)
+            state.pushNewState()
         case is BeginElse:
-            let ifState = stack.removeLast()
-            ifStack.append(ifState)
-            typeChanges += getStateChanges(oldState: ifState, newState: stack.last!)
+            state.saveIfState()
+            // Exactly one of if/else block is executed. So we should reset types from if block now.
+            state.mergeStates(typeChanges: &typeChanges, reset: true)
         case is EndIf:
-            let ifState = ifStack.removeLast()
-            let elseState = stack.removeLast()
-            stack.append(merge(ifState, elseState))
-            typeChanges += getStateChanges(oldState: elseState, newState: stack.last!)
-        case is BeginWhile:
-            stack.append(currentState)
-        case is EndWhile:
-            let loopState = stack.removeLast()
-            let previousState = stack.removeLast()
-            stack.append(merge(loopState, previousState))
-            typeChanges += getStateChanges(oldState: loopState, newState: stack.last!)
-        case is BeginDoWhile:
-            stack.append(currentState)
-        case is EndDoWhile:
-            let loopState = stack.removeLast()
-            let previousState = stack.removeLast()
-            stack.append(merge(loopState, previousState))
-            typeChanges += getStateChanges(oldState: loopState, newState: stack.last!)
-        case is BeginFor:
-            stack.append(currentState)
-        case is EndFor:
-            let loopState = stack.removeLast()
-            let previousState = stack.removeLast()
-            stack.append(merge(loopState, previousState))
-            typeChanges += getStateChanges(oldState: loopState, newState: stack.last!)
-        case is BeginForIn:
-            stack.append(currentState)
-        case is EndForIn:
-            let loopState = stack.removeLast()
-            let previousState = stack.removeLast()
-            stack.append(merge(loopState, previousState))
-            typeChanges += getStateChanges(oldState: loopState, newState: stack.last!)
-        case is BeginForOf:
-            stack.append(currentState)
-        case is EndForOf:
-            let loopState = stack.removeLast()
-            let previousState = stack.removeLast()
-            stack.append(merge(loopState, previousState))
-            typeChanges += getStateChanges(oldState: loopState, newState: stack.last!)
+            state.mergeStates(from: .conditional, typeChanges: &typeChanges)
+        case is BeginWhile, is BeginDoWhile, is BeginFor, is BeginForIn, is BeginForOf, is BeginAnyFunctionDefinition:
+            state.pushNewState()
+        case is EndWhile, is EndDoWhile, is EndFor, is EndForIn, is EndForOf, is EndAnyFunctionDefinition:
+            state.mergeStates(typeChanges: &typeChanges)
         case is BeginTry:
```

Comment:
Don't you also need to handle those to deal with variables going out of scope now?

Reply:
well yeah it was the original idea, but I kind of lost track what is the benefit of doing that. We do not want to forward this info outside interpreter and we never ask for such type inside interpreter.
So what is the point to make our life harder?

Reply:
Ok I thought that would be required to not leak type updates for dead variables into `Program.types`, but I guess if its not, then let's not do that! =)